libs/editor/egui_editor/src/bounds.rs

```diff
@@ -514,12 +553,188 @@ impl DocCharOffset {
     }
 }
 
+pub trait RangesExt {
+    /// Efficiently finds the possibly empty (inclusive, exclusive) range of ranges that contain `offset`
+    fn find_containing(
+        &self, offset: DocCharOffset, start_inclusive: bool, end_inclusive: bool,
+    ) -> (usize, usize);
+
+    /// Efficiently finds the possibly empty (inclusive, exclusive) range of ranges that intersect `range`
+    fn find_intersecting(
+        &self, range: (DocCharOffset, DocCharOffset), allow_empty: bool,
+    ) -> (usize, usize);
+}
+
+impl<Range: RangeExt<DocCharOffset>> RangesExt for Vec<Range> {
+    fn find_containing(
+        &self, offset: DocCharOffset, start_inclusive: bool, end_inclusive: bool,
+    ) -> (usize, usize) {
+        match self.binary_search_by(|range| {
+            if offset < range.start() {
+                Ordering::Less
+            } else if offset == range.start() {
+                if start_inclusive {
+                    Ordering::Equal
+                } else {
+                    Ordering::Less
+                }
+            } else if offset > range.start() && offset < range.end() {
+                Ordering::Equal
+            } else if offset == range.end() {
+                if end_inclusive {
+                    Ordering::Equal
+                } else {
+                    Ordering::Greater
+                }
+            } else if offset > range.end() {
+                Ordering::Greater
+            } else {
+                unreachable!()
+            }
+        }) {
+            Ok(idx) => {
+                let mut start = idx;
+                while start > 0 && self[idx - 1].contains(offset, start_inclusive, end_inclusive) {
+                    start -= 1;
+                }
+
+                let mut end = idx;
+                while end < self.len() - 1
+                    && self[end + 1].contains(offset, start_inclusive, end_inclusive)
+                {
+                    end += 1;
+                }
+
+                (start, end)
+            }
+            Err(idx) => (idx, idx),
+        }
+    }
+
+    fn find_intersecting(
+        &self, range: (DocCharOffset, DocCharOffset), allow_empty: bool,
+    ) -> (usize, usize) {
+        let (start_start, _) = self.find_containing(range.start(), allow_empty, false);
+        let (_, end_end) = self.find_containing(range.end(), false, allow_empty);
+        (start_start, end_end)
+    }
+}
+
+pub fn join<const N: usize>(ranges: [&[(DocCharOffset, DocCharOffset)]; N]) -> RangeJoinIter<N> {
+    let mut result = RangeJoinIter {
+        ranges,
+        in_range: [false; N],
+        current: [None; N],
+        current_end: Some(0.into()),
+    };
+    for (idx, range) in ranges.iter().enumerate() {
+        if !range.is_empty() {
+            result.current[idx] = Some(0);
+        }
+    }
+    result
+}
```

Comment:
factoring/flex: logic for iterating several bounds at once factored out of galley calculation using const generics